src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -1857,5 +1704,101 @@ private enum SendState
             /// </summary>
             Closed
         }
+
+        private interface ISendBufferAdapter
+        {
+            void SetupSendState(State state);
+
+            bool IsEmpty { get; }
+
+            static unsafe void Reserve(State state, int count)
+            {
+                if (state.SendBufferMaxCount < count)
+                {
+                    NativeMemory.Free((void*)state.SendQuicBuffers);
```

Comment:
I haven't looked much into this yet, but it feels like it might be better to wait for https://github.com/dotnet/runtime/pull/68288 to go in to be able to reuse stuff like `MsQuicBuffers` that was added there?

Reply:
Yes, please wait with this on my PR.